crates/core/src/backend/decrypt.rs

```diff
@@ -395,12 +411,41 @@ impl<C: CryptoKey> DecryptWriteBackend for DecryptBackend<C> {
             }
             None => self.key().encrypt_data(data)?,
         };
-        let id = hash(&data);
-        self.write_bytes(tpe, &id, false, data.into())
+        if self.extra_check {
+            let check_data = self.decrypt_file(&data_encrypted)?;
+            if data != check_data {
+                return Err(CryptBackendErrorKind::ExtraCheckFailed.into());
+            }
+        }
+        let id = hash(&data_encrypted);
+        self.write_bytes(tpe, &id, false, data_encrypted.into())
             .map_err(RusticErrorKind::Backend)?;
         Ok(id)
     }
 
+    fn process_data(&self, data: &[u8]) -> RusticResult<(Vec<u8>, u32, Option<NonZeroU32>)> {
+        let data_len: u32 = data
+            .len()
+            .try_into()
+            .map_err(CryptBackendErrorKind::IntConversionFailed)?;
+        let (data_encrypted, uncompressed_length) = match self.zstd {
+            None => (self.key.encrypt_data(data)?, None),
+            // compress if requested
+            Some(level) => (
+                self.key.encrypt_data(&encode_all(data, level)?)?,
+                NonZeroU32::new(data_len),
+            ),
+        };
+        if self.extra_check {
+            let data_check =
+                self.read_encrypted_from_partial(&data_encrypted, uncompressed_length)?;
+            if data != data_check {
+                return Err(CryptBackendErrorKind::ExtraCheckFailed.into());
```

Comment:
If we are doing it twice, to keep it DRY, maybe we should factor that out to its own function and `inline` it?

Also: I'm reviewing online, and need to dive into the code late, but just as a remark: `read_encrypted_from_partial` is actually return unencrypted data here to compare?

Reply:
sorry, don't see the repetition. And yes, it does return decrypted data.